### DISCUSSION TITLE
aero/coefficient properties returning 0

### DISCUSSION THREAD

**Discussion by Dobid** [JSBSim-Team/jsbsim#843] [General]
**I'm submitting a ...**
  - [x] bug report
  - [ ] feature request
  - [ ] support request => **Please do not submit support request here**, see note at the top of this template.

Hello, I'm using the c172p aircraft and I try to get the aero coefficients. But whenever I try to get them like this : 
```python
Clp = fdm['aero/coefficient/Clp']
```
I get 0.0 values. I have no other choice but to go to the .xml specification file and copy the values from it. Is there a way to get the coefficients with some python API call ? 

**Please tell us about your environment:**
 - OS : Ubuntu 20.04
 - JSBSim version : v1.1.13, installed via pip
 - Language if applicable : Python 3.11 

Thanks !

**Comment by seanmcleod** [comment]
You haven't given enough context. At what point are you executing:

```python
Clp = fdm['aero/coefficient/Clp']
```

While the aircraft is flying or while it's standing still on the ground?

Now I presume you've seen this function in `c172p.xml` right?

https://github.com/JSBSim-Team/jsbsim/blob/ebb770ebc0ec4cee6aefef6ad88702ef1164b5d7/aircraft/c172p/c172p.xml#L733-L743

Now the first thing you need to realise is that the name of the **function** can be anything that the FDM author feels like using, but typically people include coefficient names in the **function** but there is no requirement, you could name this `aaa` for example.

Now it turns out that functions using this sort of naming convention can be misleading, the `description` field gives it away, this isn't a function which returns a coefficient, it's a function that calculates and returns a **moment**, in this particular case by multiplying a number of fields and values together.

Now I'm guessing you're trying to retrieve the coefficient in particular, i.e. `-0.4840` in this example?

If so, remember that coefficients in general aren't static figures. For example take a look at this example:

https://github.com/JSBSim-Team/jsbsim/blob/ebb770ebc0ec4cee6aefef6ad88702ef1164b5d7/aircraft/737/737.xml#L770-L785

The coefficient in the above case is dependent on Mach.

What exactly are you trying to do?

**Reply by Dobid** [reply]
Got it, it doesn't return coeffs but rather moments.

> Now I'm guessing you're trying to retrieve the coefficient in particular, i.e. `-0.4840` in this example?

Yes, exactly

I'm trying to retrieve the coefficients in order to compute transfer functions coefficients for implementing a PID autopilot from the following book : Small Unmanned Aircrafts by Beard and McLain. Following the book, some transfer functions coefficients are used to help finding the right PID gains.

Side note : I'm originally working with the x8 UAV but tried to replicate my issue with the c172p to post the issue here since I assumed you'll be more familiar with the c172p. In case you want to see the fdm of the x8 : [here](https://github.com/AOS55/Fixedwing-Airsim/blob/master/x8/x8.xml).

You can see that for the x8, there's no table data for aero coefficients depending on other values, everything seems to be a constant. Can I retrieve those by calling a function ? So if I need to modify it in the fdm.xml, I don't need to update the code.

**Comment by seanmcleod** [comment]
Do you control the FDM or do you want to somehow retrieve coefficients for any FDM that you're given without editing it?

If you control the FDM what I've done in the past for my VSS (Variable Stability System) simulator is the following:

```xml
    <aerodynamics>

        <property value="-0.6">    aero/stability_coefficients/Cm-alpha </property>
        <property value="-27.0">   aero/stability_coefficients/Cm-q </property>
        <property value="-16.0">   aero/stability_coefficients/Cm-alpha-dot </property>
        <property value="1.0">     aero/stability_coefficients/Cm-d-elevator </property>

        <property value="-0.2">    aero/stability_coefficients/Cl-beta </property>
        <property value="-0.4">    aero/stability_coefficients/Cl-p </property>
        <property value="0.09">    aero/stability_coefficients/Cl-r </property>
        <property value="1.0">     aero/stability_coefficients/Cl-d-aileron </property>
        <property value="0.01">    aero/stability_coefficients/Cl-d-rudder </property>

        <property value="0.1">    aero/stability_coefficients/Cn-beta </property>
        <property value="0.00">    aero/stability_coefficients/Cn-p-xx </property>
        <property value="-0.35">   aero/stability_coefficients/Cn-r </property>
        <property value="0.00">    aero/stability_coefficients/Cn-d-aileron-xx </property>
        <property value="-0.2">    aero/stability_coefficients/Cn-d-rudder </property>

        <axis name="PITCH" unit="LBS*FT">
            <function name="aero/coefficient/Cmalpha">
                <description>Pitch_moment_due_to_alpha</description>
                <product>
                    <property>aero/qbar-psf</property>
                    <property>metrics/Sw-sqft</property>
                    <property>metrics/cbarw-ft</property>
                    <property>aero/alpha-rad</property>
                    <property>aero/stability_coefficients/Cm-alpha</property>
                </product>
            </function>
```

I'm then able to at run-time via the instructor station software read and update the coefficients to change the flying characteristics of the aircraft for a particular test pilot lesson.

**Reply by Dobid** [reply]
I mean, I suppose I'd like to retrieve the coeffs without having to edit the fdm, but since I'm mainly working on the x8 and sometimes the c172p, I don't think it is going to be too much of a hassle to modify the fdms the way you showed.
Thanks for the quick replies and solution !

**Reply by seanmcleod** [reply]
But without JSBSim enforcing some standard of how coefficients need to be supplied there is no general way to retrieve them. To JSBSim it's simply supplied with a bunch of functions which it executes on each time step, it has no specific knowledge of what the individual components of the function are.

**Comment by seanmcleod** [comment]
> implementing a PID autopilot from the following book : Small Unmanned Aircrafts by Beard and McLain

Looks like an interesting book. I presume you've come across this repo for the book?

https://github.com/randybeard/uavbook

**Reply by Dobid** [reply]
Yup I already came across this repo !
